src/lib/array.jl

```diff
@@ -331,6 +331,27 @@ end
   end
 end
 
+@adjoint eachrow(x::AbstractVecOrMat) = collect(eachrow(x)), dys -> ∇eachslice(dys, x, 1)
+@adjoint eachcol(x::AbstractVecOrMat) = collect(eachcol(x)), dys -> ∇eachslice(dys, x, 2)
```

Comment:
Why do we need to collect? Seems like it's not used in the backwards pass

Reply:
Without it, the gradient is (sometimes) some weird NamedTuple containing fields of the generator, and I got tired of trying to sort that out.

For a square matrix, it's one allocation of length N. The reverse pass is at least N such vector allocations. Master has, in addition, N matrix allocations.

Reply:
I'm aware of what master does, but this would break any laziness contract of the function which code user code could need.

Reply:
I'm doubtful, but maybe you have an example in mind? 

Obviously one could make a pathalogical `@assert x isa Generator` but I mean plausibly useful code. 

Note that anything which broadcasts already collects.

Reply:
What cases do you see the gradient for the generator in? Maybe handling that would solve more cases than without.

One could think of several bad examples, of things like lazy arrays or "infinite" arrays that are generated lazily.

Reply:
When last I tried, Zygote + LazyArrays didn't work well at all. But maybe things have improved, if you have an example in mind, please share.

The obvious use case is `reduce(hcat, map(fun, eachcol(x)))` which is a faster `mapslices`, and would now have a (fairly) fast gradient. The lazy version `reduce(hcat, fun(xi) for xi in eachcol(x))` is much slower, without mentioning Zygote.

Reply:
Last I tried it needed a couple basic adjoints around `apply` and that helped with a large chunk of it.